Fix lookup join operator not loading probe pages in order

sopel39 opened the pull request:
This is relevant only for `work_processor_pipelines`

findepi commented on presto-main/src/main/java/io/prestosql/operator/LookupJoinOperator.java:
```diff
@@ -676,18 +677,23 @@ private boolean outerJoinCurrentPosition()
 
-            boolean consumedInput = false;
-            if (needsInput() && inputPage != null) {
+            if (inputPage != null && !consumedInputPage) {
```
`inputPage != null` is known to be true

sopel39 replied:
> inputPage != null is known to be true

It isn't known. `isFinished()` might return `false` if there is pending `JoinProbe` (e.g because join yielded or output page builder was full).

findepi commented on presto-main/src/main/java/io/prestosql/operator/LookupJoinOperator.java:
```diff
@@ -676,18 +677,23 @@ private boolean outerJoinCurrentPosition()
 
             Page outputPage = getOutput();
+
+            boolean needsMoreData = consumedInputPage && needsInput();
```
`consumedInputPage` is known to be true

findepi commented on presto-main/src/main/java/io/prestosql/operator/LookupJoinOperator.java:
```diff
@@ -676,18 +677,23 @@ private boolean outerJoinCurrentPosition()
+
+            boolean needsMoreData = consumedInputPage && needsInput();
+            if (needsMoreData) {
+                consumedInputPage = false;
```
once you simplify (per above comments), do you still need `consumedInputPage` as a class state?

sopel39 commented:
AC

> A basic reading of the code it seems like the inputPage isn't always added. I assume that is always added, but it would make me more comfortable if it was obvious it always gets processed

I've added comment and simplified conditions a bit. Hopefully it's more clean now. It might not be as simple for the moment because I didn't want to modify internal join code (and complex state) just yet.

In the future, I think join complex internal flow might be simplified by modeling basic building blocs as `WorkProcessors`. For example, I imagine it could be something like:
```
(flatMap) --| inputPages ---------> join ----|------> outputPages
            | spilledPartition1 --> join ----|
            | ...                            |
            | spilledPartitionN --> join ----|
```
here `join` piece doesn't care where input pages come from and unspilling sequence is governed by `flatMap` piece that constructs `WorkProcessor` mini-pipelines for processing input pages as well as unspilled partitions.

findepi commented on presto-main/src/main/java/io/prestosql/operator/LookupJoinOperator.java:
```diff
@@ -676,18 +678,25 @@ private boolean outerJoinCurrentPosition()
+
+            boolean fetchNextInputPage = !inputFinished && needsInput();
+            if (fetchNextInputPage) {
+                addedInputPage = false;
```
+ `// reset `addedInputPage` for the sake of next `inputPage`

findepi commented:
@sopel39 i don't understand:

> Fix lookup join operator not loading probe pages in order

This says "in order", whereas the code looks like they some pages were being skipped, not reordered.
What am I missing?

sopel39 commented:
> This says "in order", whereas the code looks like they some pages were being skipped, not reordered.
> What am I missing?

Previously next `inputPage` was fetched when `addInput` was invoked on previous probe page. However, join could produce unmaterialized output page based on the previous probe page. When such output page was then materialized, the probe reader (e.g ORC) already fetched next page which resulted in out-of-bound read.

sopel39 commented:
> Much better, but the code is still tricky.

I will try to simplify join internal state.
